Handle Pipewire Release Function

dorindabassey opened the pull request:
this is based on the previous PR about Pipewire prepare

MatiasVara commented on crates/sound/src/audio_backends/pw_backend.rs:
```diff
@@ -383,4 +383,33 @@ impl AudioBackend for PwBackend {
+            println!("stream listener not found for the given stream_id");
+        };
+
+        if stream_states[stream_id as usize] == VIRTIO_SND_R_PCM_PREPARE {
```
I think the stream shall be set here in the VIRTIO_SND_R_PCM_RELEASE state. Then, in prepare(), if the stream is in RELEASE, the stream shall be created and connected again. This should be added as a OR in the if in prepare().

dorindabassey replied:
if the stream state is in Release, it disconnects and unregisters the stream, no destruction is done yet because of some issue upstream, so if a prepare is encountered in a Release(which I think is the norm), set state to set_params which ensures that the stream is created and connected again. I am not sure I understand the OR case you are refering to, do you mind to explain? thanks.

MatiasVara replied:
We can keep it like this for now but I think we shall review the PCM state-machine at some point to be conform to the spec.

stefano-garzarella commented on crates/sound/src/audio_backends/pw_backend.rs:
```diff
@@ -383,4 +382,30 @@ impl AudioBackend for PwBackend {
+        if let Some(stream) = stream_hash.get(&stream_id) {
+            stream.disconnect().expect("could not disconnect stream");
+        } else {
+            println!("stream not found for the given stream_id");
```
What about using `error!` or `warn!`?

IIUC this error is not expected so, we don't want to fail the release in the middle and continue with the next steps, right?

At this point, is the `Result<()>` the right return type, or we should force all the backend to avoid failures without returning it?

dorindabassey replied:
ok, I could use `error!`

stefano-garzarella commented on crates/sound/src/vhu_sound.rs:
```diff
@@ -381,7 +381,36 @@ impl VhostUserSoundThread {
+                        .read_obj::<VirtioSoundPcmHeader>(desc_request.addr())
+                        .map_err(|_| Error::DescriptorReadFailed)?;
+                    let stream_id = u32::from(pcm_hdr.stream_id);
+                    let state = audio_backend.get_stream_state(stream_id).unwrap();
```
Should we propagate a `get_stream_state()` error?

stefano-garzarella replied:
Or handle it in some way, like responding with `VIRTIO_SND_S_IO_ERR`?

dorindabassey replied:
I prefer the later

dorindabassey replied:
I did the former ;)

Commit:
Handle Pipewire Release function

Add functions to disconnect stream and remove stream listener, set state at SET_PARAMS if PREPARE state to
enable adding a listener and connecting the stream again.

Signed-off-by: Dorinda Bassey <dbassey@redhat.com>
